### PR TITLE
Remove suggest transport action

### PR DESCRIPTION
**Pull request by areek**
Currently `suggest` through `_search` & `_suggest` endpoints use different transport actions. 
The `_suggest` endpoint uses its own transport action to avoid executing search & aggregration 
phases. This change removes dedicated suggest transport action and adds optimization to 
shortcut to suggest phase when `_search` endpoint is used exclusively for suggestions.

This implies:

``` bash
GET /_suggest
{....}
```

is actually sugar for:

```
GET /_search
{
  "suggest" : ....
}
```

Besides reducing code and complexity of dealing with multiple execution paths for suggest, 
this change has two major benefits:
- simplifies implementing document fetching for `completion`
- allows using the search context to process suggestions
  - we can use `post_filter` / `query` for filtering suggestions in `completion`
  - `completion` can respect `_type` (through contexts)
  - simplify `phrase` suggester parameters 

**API Changes:**

This change removes `suggest` action from the client API:

``` java
client().prepareSuggest().addSuggestion("foo", ...)
```

is equivalent to:

``` java
client().prepareSearch().suggest(new SuggestBuilder().addSuggestion("foo", ...))
```

**Behaviour changes:**
now we use the `search` threadpool instead of `suggest` for suggest-only requests 
(would appreciate some feedback there) and the merged suggest stats are representative 
of all suggest-only requests either from `_suggest` or `_search` endpoints

relates to https://github.com/elastic/elasticsearch/issues/10217
simplifies https://github.com/elastic/elasticsearch/pull/13576

### CODE REVIEW

**Comment by s1monw** [comment]
+100 for the diff stats :)

**Comment by s1monw** [comment]
@areek this only breaks the java API right?

**Comment by areek** [comment]
@s1monw yes, and the `suggest` enum for the search stats request is a no-op as the suggest stats are merged with the search stats.

**Comment by s1monw** [comment]
> @s1monw yes, and the suggest enum for the search stats request is a no-op as the suggest stats are merged with the search stats.

I was more curious about the REST endpoints but I guess I will see in a second when I am reviewing :)

File: core/src/main/java/org/elasticsearch/action/search/SearchRequest.java [hunk 1]
```diff
@@ -295,6 +295,13 @@ public Boolean requestCache() {
         return this.requestCache;
     }
 
+    /**
+     * @return true if the request only has suggest
+     */
+    public boolean hasOnlySuggest() {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/search/SearchRequest.java]: can we name this `isSuggestOnly()`?

File: core/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java [hunk 2]
```diff
@@ -68,9 +71,18 @@ protected void doExecute(SearchRequest searchRequest, ActionListener<SearchRespo
             Map<String, Set<String>> routingMap = indexNameExpressionResolver.resolveSearchRouting(clusterState,
                     searchRequest.routing(), searchRequest.indices());
             int shardCount = clusterService.operationRouting().searchShardsCount(clusterState, concreteIndices, routingMap);
+            boolean hasOnlySuggest = searchRequest.hasOnlySuggest();
             if (shardCount == 1) {
                 // if we only have one group, then we always want Q_A_F, no need for DFS, and no need to do THEN since we hit one shard
                 searchRequest.searchType(QUERY_AND_FETCH);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java]: can we simplify this and have one top level block like this:

```Java

if (isSuggestOnly) {
   // disable request cache if we have only suggest
   searchRequest.requestCache(false);
  switch (searchRequest.searchType()) {
    case DFS_QUERY_AND_FETCH:
    case DFS_QUERY_THEN_FETCH:
      // disable request cache if we have only suggest
      searchRequest.searchType(QUERY_THEN_FETCH);
      break;
  }

}
```

**Comment by s1monw** [comment]
I left some minor comments - this looks fantastic! Please add the stats change to the breaking docs too.

thanks so much for this cleanup

File: core/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java (hunk 2 above)
Comment (areek) [inline core/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java]: Thanks! much cleaner :)

**Comment by areek** [comment]
Thanks @s1monw for the review! I added breaking docs for the java api and merging suggest stats into search stats. 
we don't use the `suggest` threadpool at all now (was used only by the dedicated suggest transport action). any thoughts? if we are ok with using `search` threadpool, we should remove the `suggest` threadpool all together.

**Comment by s1monw** [comment]
+1 on removing suggest threadpool

**Comment by s1monw** [comment]
LGTM - awesome lets remove the threadpool in a sep PR, we also need a breaking doc addition for that

**Comment by areek** [comment]
Thanks @s1monw for the review! I will open a PR for removing suggest threadpool

**Pull request merged**